internal/app.go

```diff
@@ -22,51 +31,115 @@ const (
 )
 
 type App struct {
-	commit  string
-	version string
+	grpcConn *grpc.GrpcConnection
+	commit   string
+	version  string
 }
 
 func NewApp(commit, version string) *App {
-	return &App{commit, version}
+	return &App{nil, version, commit}
 }
 
-func (a *App) Run(ctx context.Context) error {
+func (a *App) Run(ctx context.Context) (err error) {
+	a.handleSignal(ctx)
+
 	config.Init(a.version, a.commit)
 
 	config.RegisterRunner(func(_ *cobra.Command, _ []string) {
-		err := config.RegisterConfigFile()
+		err = config.RegisterConfigFile()
 		if err != nil {
 			slog.ErrorContext(ctx, "Failed to load configuration file", "error", err)
 			return
 		}
 
-		agentConfig, err := config.ResolveConfig()
-		if err != nil {
-			slog.ErrorContext(ctx, "Invalid config", "error", err)
+		agentConfig, configErr := config.ResolveConfig()
+		if configErr != nil {
+			slog.ErrorContext(ctx, "Invalid config", "error", configErr)
 			return
 		}
 
-		slogger := logger.New(*agentConfig.Log)
-		slog.SetDefault(slogger)
+		standardLogger := logger.New(*agentConfig.Log)
+		slog.SetDefault(standardLogger)
 
 		slog.InfoContext(ctx, "Starting NGINX Agent",
 			slog.String("version", a.version),
 			slog.String("commit", a.commit),
 		)
 
 		messagePipe := bus.NewMessagePipe(defaultMessagePipeChannelSize)
-		err = messagePipe.Register(defaultQueueSize, plugin.LoadPlugins(ctx, agentConfig))
-		if err != nil {
-			slog.ErrorContext(ctx, "Failed to register plugins", "error", err)
+		msgPipeErr := messagePipe.Register(defaultQueueSize, a.loadPlugins(agentConfig))
+		if msgPipeErr != nil {
+			slog.ErrorContext(ctx, "Failed to register plugins", "error", msgPipeErr)
 			return
 		}
 
 		messagePipe.Run(ctx)
 	})
-	err := config.Execute(ctx)
+	err = config.Execute(ctx)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func (a *App) handleSignal(ctx context.Context) {
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		select {
+		case <-sigChan:
+			slog.DebugContext(ctx, "NGINX Agent App exiting")
+		case <-ctx.Done():
+			if a.grpcConn != nil {
+				connCloseErr := a.grpcConn.Close(ctx)
+				slog.ErrorContext(ctx, "Issue closing gRPC connection", "error", connCloseErr)
```

Comment:
Should we have a nil check here for `connCloseErr` before logging that there was an error?